shared/macros/10-ansible.jinja

```diff
@@ -1022,3 +1048,350 @@ Part of the grub2_bootloader_argument_absent template.
+    authselect_current_profile: "{{ result_authselect_profile.stdout }}"
+    authselect_custom_profile: "custom/hardening"
+  when:
+    - result_authselect_profile is not skipped
+    - result_authselect_profile.stdout is not match("custom/")
+
+- name: '{{{ rule_title }}} - Get authselect current features to also enable them in the custom profile'
+  ansible.builtin.shell:
+    cmd: authselect current | tail -n+3 | awk '{ print $2 }'
+  register: result_authselect_features
+  changed_when: false
+  when:
+    - result_authselect_profile is not skipped
+    - authselect_current_profile is not match("custom/")
+
+- name: '{{{ rule_title }}} - Check if any custom profile with the same name was already created'
+  ansible.builtin.stat:
+    path: /etc/authselect/{{ authselect_custom_profile }}
+  register: result_authselect_custom_profile_present
+  changed_when: false
+  when:
+    - authselect_current_profile is not match("custom/")
+
+- name: '{{{ rule_title }}} - Create an authselect custom profile based on the current profile'
+  ansible.builtin.command:
+    cmd: authselect create-profile hardening -b {{ authselect_current_profile }}
+  when:
+    - result_authselect_check_cmd is success
+    - authselect_current_profile is not match("custom/")
+    - not result_authselect_custom_profile_present.stat.exists
+
+- name: '{{{ rule_title }}} - Ensure a backup of current authselect profile before selecting the custom profile'
+  ansible.builtin.command:
+    cmd: authselect apply-changes -b --backup=before-hardening-custom-profile.backup
+  register: result_authselect_backup
+  when:
+    - result_authselect_check_cmd is success
+    - result_authselect_profile is not skipped
+    - authselect_current_profile is not match("custom/")
+    - authselect_custom_profile is not match(authselect_current_profile)
+
+- name: '{{{ rule_title }}} - Ensure the authselect custom profile is selected'
+  ansible.builtin.command:
+    cmd: authselect select {{ authselect_custom_profile }}
+  register: result_pam_authselect_select_profile
+  when:
+    - result_authselect_check_cmd is success
+    - result_authselect_profile is not skipped
+    - authselect_current_profile is not match("custom/")
+    - authselect_custom_profile is not match(authselect_current_profile)
+
+- name: '{{{ rule_title }}} - Restore the authselect features in the custom profile'
+  ansible.builtin.command:
+    cmd: authselect enable-feature {{ item }}
+  loop: "{{ result_authselect_features.stdout_lines }}"
+  when:
+    - result_authselect_profile is not skipped
+    - result_authselect_features is not skipped
+    - result_pam_authselect_select_profile is not skipped
+
+- name: '{{{ rule_title }}} - Ensure the authselect custom profile changes are applied'
+  ansible.builtin.command:
+    cmd: authselect apply-changes -b --backup=after-hardening-custom-profile.backup
+  when:
+    - result_authselect_check_cmd is success
+    - result_authselect_profile is not skipped
```

Comment:
Can this task be made idempotent, so that it doesn't need to `apply-changes` when no custom/profile was created, nor any feature enabled?

I get the following log when running a second time, a lot of tasks `ok:` or `skipping:`, and then a task resulting in `changed:`:
```shell
TASK [Limit Password Reuse: password-auth - Get authselect current profile] **************************************************************************************************************************************************************************************************************************************************
Thursday 07 July 2022  10:22:21 +0200 (0:00:00.065)       0:02:18.121 *********                                                                                                                                                                                                                                               
ok: [rhel87n] => {"changed": false, "cmd": "authselect current -r | awk '{ print $1 }'", "delta": "0:00:00.008621", "end": "2022-07-07 04:22:21.836244", "msg": "", "rc": 0, "start": "2022-07-07 04:22:21.827623", "stderr": "", "stderr_lines": [], "stdout": "custom/hardening", "stdout_lines": ["custom/hardening"]}     
                                                                                                                                                                                                                                                                                                                              
TASK [Limit Password Reuse: password-auth - Define the current authselect profile as a local fact] ***************************************************************************************************************************************************************************************************************************
Thursday 07 July 2022  10:22:21 +0200 (0:00:00.565)       0:02:18.686 *********                                                                                                                                                                                                                                               
ok: [rhel87n] => {"ansible_facts": {"authselect_current_profile": "custom/hardening", "authselect_custom_profile": "custom/hardening"}, "changed": false}                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                                              
TASK [Limit Password Reuse: password-auth - Define the new authselect custom profile as a local fact] ************************************************************************************************************************************************************************************************************************
Thursday 07 July 2022  10:22:21 +0200 (0:00:00.075)       0:02:18.762 *********                                                                                                                                                                                                                                               
skipping: [rhel87n] => {"changed": false, "skip_reason": "Conditional result was False"}                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                                                              
TASK [Limit Password Reuse: password-auth - Get authselect current features to also enable them in the custom profile] *******************************************************************************************************************************************************************************************************
Thursday 07 July 2022  10:22:22 +0200 (0:00:00.049)       0:02:18.812 *********                                                                                                                                                                                                                                               
skipping: [rhel87n] => {"changed": false, "skip_reason": "Conditional result was False"}                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                                                              
TASK [Limit Password Reuse: password-auth - Check if any custom profile with the same name was already created] **************************************************************************************************************************************************************************************************************
Thursday 07 July 2022  10:22:22 +0200 (0:00:00.052)       0:02:18.864 *********                                                                                                                                                                                                                                               
skipping: [rhel87n] => {"changed": false, "skip_reason": "Conditional result was False"}                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                                                              
TASK [Limit Password Reuse: password-auth - Create an authselect custom profile based on the current profile] ****************************************************************************************************************************************************************************************************************
Thursday 07 July 2022  10:22:22 +0200 (0:00:00.047)       0:02:18.911 *********                                                                                                                                                                                                                                               
skipping: [rhel87n] => {"changed": false, "skip_reason": "Conditional result was False"}                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                                                              
TASK [Limit Password Reuse: password-auth - Ensure a backup of current authselect profile before selecting the custom profile] ***********************************************************************************************************************************************************************************************
Thursday 07 July 2022  10:22:22 +0200 (0:00:00.046)       0:02:18.958 *********                                                                                                                                                                                                                                               
skipping: [rhel87n] => {"changed": false, "skip_reason": "Conditional result was False"}                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                                                              
TASK [Limit Password Reuse: password-auth - Ensure the authselect custom profile is selected] ********************************************************************************************************************************************************************************************************************************
Thursday 07 July 2022  10:22:22 +0200 (0:00:00.058)       0:02:19.017 *********                                                                                                                                                                                                                                               
skipping: [rhel87n] => {"changed": false, "skip_reason": "Conditional result was False"}                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                                                              
TASK [Limit Password Reuse: password-auth - Restore the authselect features in the custom profile] ***************************************************************************************************************************************************************************************************************************
Thursday 07 July 2022  10:22:22 +0200 (0:00:00.056)       0:02:19.073 *********                                                                                                                                                                                                                                               
skipping: [rhel87n] => {"changed": false, "skip_reason": "Conditional result was False"}                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                                                              
TASK [Limit Password Reuse: password-auth - Ensure the authselect custom profile changes are applied] ************************************************************************************************************************************************************************************************************************
Thursday 07 July 2022  10:22:22 +0200 (0:00:00.185)       0:02:19.258 *********                                                                                                                                                                                                                                               
changed: [rhel87n] => {"changed": true, "cmd": ["authselect", "apply-changes", "-b", "--backup=after-hardening-custom-profile.backup"], "delta": "0:00:00.025319", "end": "2022-07-07 04:22:22.984635", "msg": "", "rc": 0, "start": "2022-07-07 04:22:22.959316", "stderr": "", "stderr_lines": [], "stdout": "Backup stored 
at /var/lib/authselect/backups/after-hardening-custom-profile.backup\nChanges were successfully applied.", "stdout_lines": ["Backup stored at /var/lib/authselect/backups/after-hardening-custom-profile.backup", "Changes were successfully applied."]}                                                                      
                                                                                                                                                                                                                                                                                                                              
TASK [Limit Password Reuse: password-auth - Change the PAM file to be edited according to the custom authselect profile] *****************************************************************************************************************************************************************************************************
Thursday 07 July 2022  10:22:23 +0200 (0:00:00.565)       0:02:19.824 *********                                                                                                                                                                                                                                               
ok: [rhel87n] => {"ansible_facts": {"pam_file_path": "/etc/authselect/custom/hardening/password-auth"}, "changed": false}                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                                              
TASK [Limit Password Reuse: password-auth - Ensure a backup of current authselect profile before changing the custom profile] ************************************************************************************************************************************************************************************************
Thursday 07 July 2022  10:22:23 +0200 (0:00:00.064)       0:02:19.888 *********                                                                                                                                                                                                                                               
changed: [rhel87n] => {"changed": true, "cmd": ["authselect", "apply-changes", "-b", "--backup=before-hardening--.backup"], "delta": "0:00:00.026537", "end": "2022-07-07 04:22:23.566245", "msg": "", "rc": 0, "start": "2022-07-07 04:22:23.539708", "stderr": "", "stderr_lines": [], "stdout": "Backup stored at /var/lib/
authselect/backups/before-hardening--.backup\nChanges were successfully applied.", "stdout_lines": ["Backup stored at /var/lib/authselect/backups/before-hardening--.backup", "Changes were successfully applied."]}                                                                                                          
                                                                                                                                                                                                                                                                                                                              
TASK [Limit Password Reuse: password-auth - Check if expected PAM module line is present in /etc/authselect/custom/hardening/password-auth] **********************************************************************************************************************************************************************************
Thursday 07 July 2022  10:22:23 +0200 (0:00:00.528)       0:02:20.416 *********                                                                                                                                                                                                                                               
ok: [rhel87n] => {"backup": "", "changed": false, "found": 1, "msg": "1 line(s) removed"}                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                                              
TASK [Limit Password Reuse: password-auth - Check if required PAM module line is present in /etc/authselect/custom/hardening/password-auth with different control] ***********************************************************************************************************************************************************
Thursday 07 July 2022  10:22:24 +0200 (0:00:00.673)       0:02:21.089 *********                                                                                                                                                                                                                                               
skipping: [rhel87n] => {"changed": false, "skip_reason": "Conditional result was False"}                                                                                                                                                                                                                                      
                                                                                                                              
```

Reply:
There are other tasks that also `authselect apply-changes`, it would be great if they could also be made idempotent.

Reply:
Hum, good catch. That one in the `ansible_ensure_authselect_custom_profile` is possible. The others I noticed during the tests and didn't find a reliable and simple way since they depend on command output which doesn't differentiate a command that actually changed or not something in a successful result.

Reply:
However, along the development and after some refinement triggered by the reviews, it seems it is doable now. Let me check again.